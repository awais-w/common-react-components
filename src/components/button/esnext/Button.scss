@import '@cmc/common-styles/src/scss/_variables.scss';

/*
 *  NOTE: expanding the selector tree from .cmc-button to .cmc-button.button
 *  so that the button class styling doesnt override the .cmc-button styling
 */

$darken-volume: 10%;

@keyframes spCircRot {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(359deg);
  }
}

.cmc-button {
  padding: 0 $super-double;
  &.button {
    background-color: $primary-color;
    border-color: $primary-color;
    display: flex;
    align-items: center;
    justify-content: center;
    height: $input-height-regular;
    position: relative;
    border-radius: $half;
    &:hover,
    &:focus,
    &:active {
      background-color: darken($primary-color, $darken-volume);
      border-color: darken($primary-color, $darken-volume);
    }
    &--secondary {
      background-color: $color-white;
      border-color: $primary-color;
      color: $primary-color;
      &:hover,
      &:focus {
        background-color: $color-white;
        border-color: darken($primary-color, $darken-volume);
        color: darken($primary-color, $darken-volume);
      }
      .cmc-button__loading {
        border-color: rgba($primary-color, 0.3);
        border-top-color: $primary-color;
      }
      &.cmc-button__icon--splitLeft,
      &.cmc-button__icon--splitRight {
        .cmc-icon-label-value .cmc-icon {
          border-color: $primary-color;
          border-width: 2px;
        }
      }
    }
    .cmc-icon-label-value {
      flex-wrap: nowrap;
      &__label {
        @include font-compressed-bold;
        font-size: $font-size-large;
        margin: 0;
      }
      .cmc-icon {
        margin: 0;
        align-self: center;
      }
    }
  }

  &__loading {
    width: $large;
    height: $large;
    border: $half rgba($color-white, 0.25) solid;
    border-top-color: rgba($color-white, 1);
    border-radius: 50%;
    animation: spCircRot 0.6s infinite linear;
    margin: auto;
    &--withLabel {
      display: inline-block;
      margin: 0 $base -$half 0;
    }
  }

  &__icon {
    .cmc-icon {
      font-size: $font-size-large;
    }
    &--left {
      .cmc-icon-label-value {
        .cmc-icon + .cmc-icon-label-value__label {
          margin-left: $base;
        }
      }
    }
    &--right {
      .cmc-icon-label-value {
        flex-direction: row-reverse;
        .cmc-icon-label-value__label {
          margin-right: $base;
          margin-left: 0;
        }
      }
    }
    &--splitLeft,
    &--splitRight {
      .cmc-icon-label-value .cmc-icon {
        position: absolute;
        height: $input-height-regular;
        display: flex;
        align-items: center;
        padding: 0 $base;
      }
    }
    &--splitLeft {
      .cmc-icon-label-value {
        margin-left: $base * 5;
        .cmc-icon {
          left: 0;
          border-right: 1px solid $color-white;
        }
      }
    }
    &--splitRight {
      .cmc-icon-label-value {
        margin-right: $base * 5;
        .cmc-icon {
          right: 0;
          border-left: 1px solid $color-white;
        }
      }
    }
  }

  &.button--contentFit {
    padding-left: $double;
    padding-right: $double;
  }
  &.button--containerFit {
    width: 100%;
    .cmc-icon-label-value {
      display: inline-flex;
    }
  }
  &.button--short {
    height: $input-height-short;
    padding-top: 0;
    padding-bottom: 0;
    .cmc-icon {
      line-height: 120%;
    }
  }
}

@include touch-device-bp1024 {
  .cmc-button.button {
    &.button--short {
      height: $input-height-regular;
    }
  }
}

@include tablet-and-up {
  .cmc-button {
    padding: 0 $super-large;
  }
  .button--half {
    padding: 0 $super-double;
  }
}
