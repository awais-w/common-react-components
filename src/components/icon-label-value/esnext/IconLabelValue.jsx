import React from 'react';
import PropTypes, { oneOf } from 'prop-types';
import Icon from '@cmc/icon/esnext/Icon';
import bem from '@argos/utils/esnext/bem';
import isEmpty from '@cmc/utils/esnext/isEmpty';
import ReactTooltip from 'react-tooltip';
import uuid from 'uuid';
import './IconLabelValue.scss';

const cn = bem({ prefix: 'cmc-', block: 'icon-label-value' });

const IconLabelValue = ({ value, icon, isCustom, isInline, label, onClick, tooltipValue, color, labelColor }) => {
  const handleOnClick = (e, index) => {
    // Potentially an onClick on each value item instead of the top level.
    // This if statement helps to handle that, ensuring the correct one is clicked.
    if (!isEmpty(index) && value && value[index] && value[index].onClick) {
      value[index].onClick(e);
    } else if (onClick) {
      onClick(e);
    }
  };

  const valueContent = !isEmpty(value) ? (
    !Array.isArray(value) ? (
      <span className={cn('value', labelColor)}>{value}</span>
    ) : (
      value.map((valueObj, index) => (
        <span
          key={valueObj.key}
          className={cn('value-multi', valueObj.onClick && 'clickable', labelColor)}
          onClick={(e) => handleOnClick(e, index)}>
          {valueObj.value}
        </span>
      ))
    )
  ) : null;

  const tooltipData = tooltipValue
    ? {
        'data-for': uuid.v4(),
        'data-tip': tooltipValue,
      }
    : {};

  return (
    <div
      className={cn(null, { inline: isInline, 'no-icon': isEmpty(icon) })}
      {...tooltipData}
      onClick={(e) => handleOnClick(e)}>
      {!isEmpty(icon) && <Icon isCustom={isCustom} icon={icon} color={color} />}
      {!isEmpty(label) && <span className={cn('label', labelColor)}>{label}</span>}
      {valueContent && valueContent}
      {tooltipValue && <ReactTooltip id={tooltipData['data-for']} effect='solid' />}
    </div>
  );
};

IconLabelValue.propTypes = {
  icon: PropTypes.string,
  label: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired,
        onClick: PropTypes.func,
      }),
    ),
  ]),
  onClick: PropTypes.func,
  isInline: PropTypes.bool,
  isCustom: PropTypes.bool,
  tooltipValue: PropTypes.string,
  labelColor: oneOf(['amber', 'blue', 'red', 'green', 'grey']),
};
export default IconLabelValue;
