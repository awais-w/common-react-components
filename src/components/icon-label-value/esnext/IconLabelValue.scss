@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-icon-label-value {
  @include font-standard;
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  column-gap: $half;

  .material-icons {
    padding-right: $half;
    user-select: none;
  }

  &--inline {
    margin-right: $base;
    display: inline-flex;

    // double-click selection fix for inline-blocks
    &::after {
      content: '';
      display: block;
    }
  }

  &--no-icon {
    &::before {
      content: '';
      height: $large;
      margin-left: -$half;
    }
  }

  &__label {
    @include font-bold;
  }

  &__label,
  &__value {
    line-height: initial;
  }

  &__value-multi {
    &--clickable {
      cursor: pointer;
      color: $color-blue-cobalt;
      text-decoration: underline;

      &:hover {
        color: $primary-color-light;
      }
    }
  }

  &__label,
  &__value {
    &--amber,
    &.amber {
      color: $color-sainsburys;
    }
    &--blue,
    &.blue {
      color: $primary-color;
    }
    &--red,
    &.red {
      color: $color-rust;
    }
    &--green,
    &.green {
      color: $color-green;
    }
    &--grey,
    &.grey {
      color: $color-grey-dark;
    }
  }
}
