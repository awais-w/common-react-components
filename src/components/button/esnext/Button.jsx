import React from 'react';
import PropTypes from 'prop-types';
import IconLabelValue from '@cmc/icon-label-value/esnext/IconLabelValue';
import bem from '@argos/utils/esnext/bem';
import './Button.scss';

const cn = bem({ prefix: 'cmc-', block: 'button' });

const iconModes = ['left', 'right', 'splitLeft', 'splitRight'];
const buttonWidths = ['default', 'half', 'contentFit', 'containerFit'];

const propTypes = {
  text: PropTypes.string,
  loadingText: PropTypes.string,
  icon: PropTypes.string,
  iconMode: PropTypes.oneOf(iconModes),
  buttonWidth: PropTypes.oneOf(buttonWidths),
  name: PropTypes.string,
  value: PropTypes.string,
  type: PropTypes.string,
  onClick: PropTypes.func,
  isDisabled: PropTypes.bool,
  isLoading: PropTypes.bool,
  isSecondary: PropTypes.bool,
  isShort: PropTypes.bool,
};

const defaultProps = {
  text: '',
  loadingText: '',
  icon: '',
  iconMode: 'left',
  buttonWidth: 'default',
  type: 'button',
  onClick: null,
  isDisabled: false,
  isLoading: false,
  isSecondary: false,
  isShort: false,
};

const Button = ({
  name,
  value,
  onClick,
  isSecondary,
  type,
  isDisabled,
  isShort,
  buttonWidth,
  icon,
  iconMode,
  isLoading,
  text,
  loadingText,
  noDiv,
}) => {
  const loadingIcon = loadingText ? (
    <React.Fragment>
      <div className={cn('loading', 'withLabel')} />
      <IconLabelValue icon={iconMode === 'left' && icon} label={loadingText} isInline={true} />
    </React.Fragment>
  ) : (
    <div className={cn('loading')} />
  );

  const buttonContent = <IconLabelValue icon={icon} label={text} />;

  const buttonClassName = cn(
    null,
    null,
    `button ${isSecondary ? 'button--secondary' : ''}
      ${isShort ? 'button--short' : ''}
      ${cn('icon', iconMode)}
      ${buttonWidth ? `button--${buttonWidth}` : ''}`,
  );

  const btn = (
    <React.Fragment>
      <button
        /*
         * NOTE: Classes added are inherited from Bolt and are not BEM modifiers of cmc-button component hence they are
         * not put in the modifier section of the cn() function.
         */
        className={buttonClassName}
        onClick={onClick}
        name={name}
        value={value}
        type={type}
        disabled={isDisabled}>
        {isLoading ? loadingIcon : buttonContent}
      </button>
    </React.Fragment>
  );
  return noDiv ? btn : <div>{btn}</div>;
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;
export default Button;
