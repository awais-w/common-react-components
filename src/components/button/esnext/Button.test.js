import React from 'react';
import { shallow } from 'enzyme';
import IconLabelValue from '@cmc/icon-label-value/esnext/IconLabelValue';
import Button from './Button';

const setup = (props) => {
  const component = shallow(<Button {...props} />);

  return {
    wrapper: component,
    getButton: () => component.find('.cmc-button'),
    getButtonType: () => component.find('.cmc-button').prop('type'),
    isSecondary: () => component.find('.button--secondary'),
    isShort: () => component.find('.button--short'),
    isDisabled: () => component.find('.cmc-button').prop('disabled'),
    getName: () => component.find('.cmc-button').prop('name'),
    getValue: () => component.find('.cmc-button').prop('value'),
    getText: () => component.find(IconLabelValue).prop('label'),
    getOnClick: () => component.find('.cmc-button').prop('onClick'),
    isLoading: () => component.find('.cmc-button__loading'),
    loadingText: () => component.find('.cmc-button__loading--withLabel'),
    splitLeftIcon: () => component.find('.cmc-button__icon--splitLeft'),
    splitRightIcon: () => component.find('.cmc-button__icon--splitRight'),
    halfWidth: () => component.find('.button--half'),
    contentWidth: () => component.find('.button--contentFit'),
    containerWidth: () => component.find('.button--containerFit'),
  };
};

describe('Button', () => {
  describe('with no props', () => {
    const component = setup();

    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });

    it('should render a button', () => {
      expect(component.getButton()).toExist();
    });

    it('should be of type button', () => {
      expect(component.getButtonType()).toBe('button');
    });
  });

  describe('with isSecondary prop', () => {
    const component = setup({ isSecondary: true });

    it('should render a secondary button', () => {
      expect(component.isSecondary()).toExist();
    });
  });

  describe('with isShort prop', () => {
    const component = setup({ isShort: true });

    it('should render a short height button', () => {
      expect(component.isShort()).toExist();
    });
  });

  describe('with isDisabled prop', () => {
    const component = setup({ isDisabled: true });

    it('should render a disabled button', () => {
      expect(component.isDisabled()).toBe(true);
    });
  });

  describe('with name prop', () => {
    const nameProp = 'name';
    const component = setup({ name: nameProp });

    it('should render a name in the button', () => {
      expect(component.getName()).toEqual(nameProp);
    });
  });

  describe('with value prop', () => {
    const valueProp = 'value';
    const component = setup({ value: valueProp });

    it('should render value in the button', () => {
      expect(component.getValue()).toEqual(valueProp);
    });
  });

  describe('with text prop', () => {
    const textProp = 'Hi I am a text prop';
    const component = setup({ text: textProp });

    it('should render text in the button', () => {
      expect(component.getText()).toEqual(textProp);
    });
  });

  describe('with onClick prop', () => {
    const onClickProp = jest.fn();
    const component = setup({ onClick: onClickProp });

    it('should render text in the button', () => {
      expect(component.getOnClick()).toEqual(onClickProp);
    });
  });

  describe('with isLoading prop', () => {
    const component = setup({ isLoading: true });

    it('should render loading button', () => {
      expect(component.isLoading()).toExist();
    });
  });

  describe('with isLoading and and loadingText prop', () => {
    const component = setup({ isLoading: true, loadingText: 'Retrying...' });

    it('should render loading button with text', () => {
      expect(component.isLoading()).toExist();
    });
  });

  describe('with split left icon', () => {
    const props = { icon: 'cancel', iconMode: 'splitLeft', onClick: jest.fn() };
    const component = setup(props);

    it('should render the split left icon', () => {
      expect(component.splitLeftIcon()).toExist();
    });
  });

  describe('with split right icon', () => {
    const props = { icon: 'cancel', iconMode: 'splitRight', onClick: jest.fn() };
    const component = setup(props);

    it('should render the split right icon', () => {
      expect(component.splitRightIcon()).toExist();
    });
  });

  describe('with half width button', () => {
    const component = setup({ buttonWidth: 'half' });

    it('should render the half button', () => {
      expect(component.halfWidth()).toExist();
    });
  });

  describe('with content width button', () => {
    const component = setup({ buttonWidth: 'contentFit' });

    it('should render the content width button', () => {
      expect(component.contentWidth()).toExist();
    });
  });

  describe('with container width button', () => {
    const component = setup({ buttonWidth: 'containerFit' });

    it('should render the container width button', () => {
      expect(component.containerWidth()).toExist();
    });
  });
});
