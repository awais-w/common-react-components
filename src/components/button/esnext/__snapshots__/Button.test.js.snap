// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button with no props should match default snapshot 1`] = `
<div>
  <button
    className="cmc-button button 
      
      cmc-button__icon cmc-button__icon--left
      button--default"
    disabled={false}
    onClick={null}
    type="button"
  >
    <IconLabelValue
      icon=""
      label=""
    />
  </button>
</div>
`;
